### PR TITLE
ShopKeeper NPC

### PR DESCRIPTION
**Pull request by dedztbh**
- invincible
- does not move on its own
- trade disabled
- callback when player interact
- /npc <name> spawns a shopkeeper NPC at player location with name <name>
- /cleannpc clears all shopkeepers

### CODE REVIEW

**Comment by Dummyc0m** [comment]
addPotionEffect(PotionEffect(PotionEffectType.SLOW, Int.MAX_VALUE, 65536, false))

LMFAO

File: src/main/kotlin/bedwars/shop/ShopKeeper.kt [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+            setAdult()
+            ageLock = true
+            profession = Villager.Profession.FARMER
+            addPotionEffect(PotionEffect(PotionEffectType.SLOW, Int.MAX_VALUE, 65536, false))
+            ShopKeeperManager.shopKeepers.add(this@ShopKeeper)
+        }
+    }
+
+    // Despawn the ShopKeeper
+    fun despawn(removeFromManager: Boolean = true) {
```
Comment (Dummyc0m) [inline src/main/kotlin/bedwars/shop/ShopKeeper.kt]: When would removeFromManager be false?

File: src/main/kotlin/bedwars/shop/ShopKeeper.kt [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        if (removeFromManager)
+            ShopKeeperManager.shopKeepers.remove(this)
+        shopKeeperEntity?.remove()
+        shopKeeperEntity = null
+    }
+}
+
+object ShopKeeperManager : Listener {
```
Comment (Dummyc0m) [inline src/main/kotlin/bedwars/shop/ShopKeeper.kt]: Can we have an event handler for when an NPC dies? If that’s possible at all. Also, rn the NPC is still technically moveable by things like water/piston, right?

File: src/main/kotlin/bedwars/shop/ShopKeeper.kt [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    fun spawn() = location.run {
+        if (shopKeeperEntity != null) return@run
+        shopKeeperEntity = world.spawn(location, Villager::class.java)?.apply {
+            customName = this@ShopKeeper.name
+            isCustomNameVisible = true
+            setAdult()
+            ageLock = true
+            profession = Villager.Profession.FARMER
+            addPotionEffect(PotionEffect(PotionEffectType.SLOW, Int.MAX_VALUE, 65536, false))
+            ShopKeeperManager.shopKeepers.add(this@ShopKeeper)
```
Comment (Dummyc0m) [inline src/main/kotlin/bedwars/shop/ShopKeeper.kt]: IMO spawn and despawn should be methods of ShopKeeperManager, then you can have ShopKeeper be a data class.

File: src/main/kotlin/bedwars/shop/ShopKeeper.kt [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    fun entityDamage(evt: EntityDamageEvent) =
+            shopKeepers.any { it.shopKeeperEntity === evt.entity } then { evt.isCancelled = true }
+
+    @EventHandler
+    fun entityDamage(evt: EntityDamageByEntityEvent) = entityDamage(evt as EntityDamageEvent)
+
+    // Disable Trading
+    @EventHandler
+    fun entityInteract(evt: PlayerInteractEntityEvent) =
+            shopKeepers.firstOrNull { it.shopKeeperEntity === evt.rightClicked }?.let {
```
Comment (Dummyc0m) [inline src/main/kotlin/bedwars/shop/ShopKeeper.kt]: I suppose this is fine? But a mutableSet is backed by a hashset, at which point you might as well use a mutableMap from UUID to shopkeeper

**Review by Dummyc0m — CHANGES_REQUESTED** [review]

File: src/main/kotlin/bedwars/protection/BlockProtection.kt [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        return Vec3(x, y, z)
+    }
+
+    private fun <T> processEvent(evt: T): Boolean where T : BlockEvent, T : Cancellable =
+            processBlock(evt.block).also {
+                if (it) evt.isCancelled = true
+            }
+
+    private fun processBlock(b: Block): Boolean =
+            protectedAreas.any {
```
Comment (buhiroshi0205) [inline src/main/kotlin/bedwars/protection/BlockProtection.kt]: I think we can add spherical region protection later

File: src/main/kotlin/bedwars/shop/ShopKeeper.kt (hunk 3 above)
Comment (buhiroshi0205) [reply src/main/kotlin/bedwars/shop/ShopKeeper.kt]: +1

File: src/main/kotlin/bedwars/shop/ShopKeeper.kt (hunk 1 above)
Comment (buhiroshi0205) [reply src/main/kotlin/bedwars/shop/ShopKeeper.kt]: If despawn was in ShopManager I don't think we would even need a boolean argument here.

File: src/main/kotlin/bedwars/shop/ShopKeeper.kt (hunk 2 above)
Comment (buhiroshi0205) [reply src/main/kotlin/bedwars/shop/ShopKeeper.kt]: Shouldn't be a big issue? I don't think this needs fixing as long as we intend to use this plugin only internally

**Review by buhiroshi0205 — CHANGES_REQUESTED** [review]

**Comment by buhiroshi0205** [comment]
LMAO first time using git pull request in an actual meaningful way, idek if I used it correctly

**Comment by buhiroshi0205** [comment]
I'll resolve the conflicts (those appeared after I updated my master branch)

**Comment by Dummyc0m** [comment]
Sounds good. Let’s try it out.

**Comment by dedztbh** [comment]
Fixed

**Review by Dummyc0m — APPROVED** [review]
looks pretty good now.
ShopKeeper as a data class has a nullable var... not ideal but whatever, fits the project

**Pull request merged**